python/lsst/daf/butler/registry/queries/_query.py

```diff
@@ -61,6 +63,222 @@
     from ._builder import QueryBuilder
 
 
+ColumnIterable = Iterable[sqlalchemy.sql.ColumnElement]
+
+
+class _OrderByHandler:
+    """Class that knows how to transform order_by list into other formats.
+
+    Parameters
+    ----------
+    order_by : `Iterable` [ `str` ]
+        Sequence of names to use for ordering with optional "-" prefix.
+    graph : `DimensionGraph`
+        Dimensions used by a query.
+    db : `Database`
+        Database instance.
+    """
+    def __init__(self, order_by: Iterable[str], graph: DimensionGraph, db: Database):
+        self._order_by = tuple(self._parse_name(name, graph) for name in order_by)
+        self._graph = graph
+        self._db = db
+
+    @staticmethod
+    def _parse_name(name: str, graph: DimensionGraph) -> Tuple[Dimension, Optional[str], bool]:
+        """Convert ordering name a dimension, field name, name and ordering.
+
+        Parameters
+        ----------
+        name : `str`
+            Name with an optional "-" prefix, name cannot be empty, it can
+            consist of dimension name and field name separated by dot.
+        graph : `DimensionGraph`
+            Dimensions that are present in a query.
+
+        Returns
+        -------
+        dimension : `Dimension`
+            `Dimension` corresponding to a name.
+        field : `str` or `None`
+            Name of the field, None if the name refers to dimension itself or
+            its primary key.
+        ascending : `bool`
+            True if order should be ascending (no "-" prefix present), False
+            otherwise.
+        """
+        if not name or name == "-":
+            raise ValueError("Empty dimension name in ORDER BY")
+        ascending = True
+        if name[0] == "-":
+            ascending = False
+            name = name[1:]
+
+        dimension: Dimension
+        field_name: Optional[str] = None
+        if "." not in name:
+            # No dot, can be either a dimension name or field name (in any of
+            # the known dimensions)
+            if name in graph.dimensions.names:
+                dimension = graph.dimensions[name]
+            elif name in ("timespan_begin", "timespan_end"):
+                matches = [dim for dim in graph if dim.temporal]
```

Comment:
```suggestion
                matches = [element for element in graph.elements if element.temporal]
```
We want this to include `DimensionElement`s that are not `Dimension`s.

Reply:
I thought I would simplify life for myself if I only use Dimensions, Elements don't have primary/alternative keys and elements that we have defined today don't have any metadata. Is it possible we'll have Elements with metadata?

Reply:
I re-worked the code to handle DimensionElements correctly.